hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/SnapshotDeletingService.java

```diff
@@ -0,0 +1,378 @@
+              toNextDBList.add(toNextDb.build());
+
+            }
+            // Submit Move request to OM.
+            submitSnapshotMoveDeletedKeys(snapInfo, nextSnapshot,
+                toActiveDBList, toNextDBList);
+            snapshotLimit--;
+          } catch (IOException ex) {
+            LOG.error("Error while running Snapshot Deleting Service", ex);
+          }
+        }
+
+        successRunCount.incrementAndGet();
+      } catch (IOException e) {
+        LOG.error("Error while running Snapshot Deleting Service", e);
+      }
+
+      return BackgroundTaskResult.EmptyTaskResult.newResult();
+    }
+
+    private void splitRepeatedOmKeyInfo(KeyValuePair.Builder toActiveDb,
+        KeyValuePair.Builder toNextDb, OmKeyInfo keyInfo,
+        Table<String, OmKeyInfo> previousKeyTable) throws IOException {
+      if (checkKeyExistInPreviousTable(previousKeyTable, keyInfo)) {
+        // Move to next non deleted snapshot's deleted table
+        toNextDb.addKeyInfos(keyInfo.getProtobuf(
+            ClientVersion.CURRENT_VERSION));
+      } else {
+        // Move to active DB Deleted Table.
+        toActiveDb.addKeyInfos(keyInfo
+            .getProtobuf(ClientVersion.CURRENT_VERSION));
+      }
+    }
+
+    private void submitSnapshotMoveDeletedKeys(SnapshotInfo snapInfo,
+        SnapshotInfo nextSnapshot, List<KeyValuePair> toActiveDBList,
+        List<KeyValuePair> toNextDBList) {
+
+      SnapshotMoveDeletedKeysRequest.Builder moveDeletedKeysBuilder =
+          SnapshotMoveDeletedKeysRequest.newBuilder()
+              .setFromSnapshot(snapInfo.getProtobuf());
+
+      if (nextSnapshot != null) {
+        moveDeletedKeysBuilder.setNextSnapshot(nextSnapshot.getProtobuf());
+      }
+
+      SnapshotMoveDeletedKeysRequest moveDeletedKeys =
+          moveDeletedKeysBuilder.addAllActiveDBKeys(toActiveDBList)
+          .addAllNextDBKeys(toNextDBList).build();
+
+
+      OMRequest omRequest = OMRequest.newBuilder()
+          .setCmdType(Type.SnapshotMoveDeletedKeys)
+          .setSnapshotMoveDeletedKeysRequest(moveDeletedKeys)
+          .setClientId(clientId.toString())
+          .build();
+
+      submitRequest(omRequest);
+    }
+
+    private boolean checkKeyExistInPreviousTable(
+        Table<String, OmKeyInfo> previousKeyTable, OmKeyInfo deletedKeyInfo)
+        throws IOException {
+
+      if (previousKeyTable == null) {
+        return false;
+      }
+
+      //TODO: Handle Renamed Keys
+      String dbKey = ozoneManager.getMetadataManager()
+          .getOzoneKey(deletedKeyInfo.getVolumeName(),
+              deletedKeyInfo.getBucketName(), deletedKeyInfo.getKeyName());
+
+      OmKeyInfo prevKeyInfo = previousKeyTable.get(dbKey);
+      if (prevKeyInfo != null &&
+          prevKeyInfo.getObjectID() == deletedKeyInfo.getObjectID()) {
+        return true;
+      }
+      return false;
+    }
+
+    private SnapshotInfo getPreviousSnapshot(SnapshotInfo snapInfo)
```

Comment:
Doesn't this method need to be bucket specfic?  See section 4.1.b.a of the design doc:

https://www.notion.so/fossil-i/Design-Ozone-Snapshot-Deletion-Garbage-Collection-based-on-key-deletedTable-2a624480dc7c4bc3ad608cbf86a25541?pvs=4#cd47388bf64142809592ebe544f64aa0

Reply:
It iterates on path based chain which is same as bucket specific.

Reply:
sounds good, thanks!